pkg/cache/inmemory.go

```diff
@@ -0,0 +1,302 @@
+		Name:        "thanos_store_cache_hits_total",
+		Help:        "Total number of requests to the inmemory cache that were a hit.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.current = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name:        "thanos_store_cache_items",
+		Help:        "Current number of items in the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.currentSize = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name: "thanos_store_cache_items_size_bytes",
+		Help: "Current byte size of items in the inmemory cache.",
+	})
+
+	c.totalCurrentSize = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name: "thanos_store_cache_total_size_bytes",
+		Help: "Current byte size of items (both value and key) in the inmemory cache.",
+	})
+
+	_ = promauto.With(reg).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "thanos_store_cache_max_size_bytes",
+		Help: "Maximum number of bytes to be held in the inmemory cache.",
+	}, func() float64 {
+		return float64(c.maxSizeBytes)
+	})
+	_ = promauto.With(reg).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "thanos_store_cache_max_item_size_bytes",
+		Help: "Maximum number of bytes for single entry to be held in the inmemory cache.",
+	}, func() float64 {
+		return float64(c.maxItemSizeBytes)
+	})
+
+	// Initialize LRU cache with a high size limit since we will manage evictions ourselves
+	// based on stored size using `RemoveOldest` method.
+	l, err := lru.NewLRU(maxInt, c.onEvict)
+	if err != nil {
+		return nil, err
+	}
+	c.lru = l
+
+	level.Info(logger).Log(
+		"msg", "created in-memory inmemory cache",
+		"maxItemSizeBytes", c.maxItemSizeBytes,
+		"maxSizeBytes", c.maxSizeBytes,
+		"maxItems", "maxInt",
+	)
+	return c, nil
+}
+
+func (c *InMemoryCache) onEvict(key, val interface{}) {
+	keySize := keyHeaderSize + uint64(len(key.(string)))
+	entrySize := sliceHeaderSize + uint64(len(val.(cacheDataWithTTLWrapper).data)) + uint64(unsafe.Sizeof(val.(cacheDataWithTTLWrapper).expiryTime))
+
+	c.evicted.Inc()
+	c.current.Dec()
+	c.currentSize.Sub(float64(entrySize))
+	c.totalCurrentSize.Sub(float64(keySize + entrySize))
+
+	c.curSize -= entrySize
+}
+
+func (c *InMemoryCache) get(key string) ([]byte, bool) {
+	c.requests.Inc()
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	v, ok := c.lru.Get(key)
+	if !ok {
+		return nil, false
+	}
+	// If the present time is greater than the TTL for the object from cache, the object will be
+	// removed from the cache and a nil will be returned
+	if time.Now().After(v.(cacheDataWithTTLWrapper).expiryTime) {
+		c.hitsExpired.Inc()
+		c.lru.Remove(key)
+		return nil, false
+	}
+	c.hits.Inc()
+	return v.(cacheDataWithTTLWrapper).data, true
+}
+
+func (c *InMemoryCache) set(key string, val []byte, ttl time.Duration) {
+	var size = sliceHeaderSize + uint64(len(val))
+	keySize := keyHeaderSize + uint64(len(key))
+	entrySize := sliceHeaderSize + uint64(len(val)) + uint64(unsafe.Sizeof(time.Now().Add(ttl)))
```

Comment:
I see.
Then I guess it makes sense to just have `size = len(val)`. The `key` is just a `String` anyway...
This would make it similar to: (without the `*HeaderSize`)
https://github.com/thanos-io/thanos/blob/b5e7b2b493783d89933874293e5c6a97a5b444b5/pkg/store/cache/inmemory.go#L207-L208

Reply:
I wouldn't skip `key`. Eg. when caching attributes via caching bucket, key may be longer than value.

Reply:
Oh okay. I think the `key` size can be summed up like this:
https://github.com/thanos-io/thanos/blob/b5e7b2b493783d89933874293e5c6a97a5b444b5/pkg/store/cache/inmemory.go#L230